### PR TITLE
feat(agent): decompose loop behind observer seam + reframe default persona

### PR DESCRIPTION
**Pull request by lvndry**
Two related harness improvements, rebased onto current `main`.

## 1. Agent-loop decomposition + observer seam (#3)

Decomposes the ~400-line `executeAgentLoop` into `runIteration` / `handleToolPhase` / `finalizeRun` (shared `LoopState`), and routes the loop's five lifecycle events (thinking / interrupted / iteration-limit / empty-response / completion) through a new **`AgentLoopObserver`** whose default adapter forwards 1:1 to `PresentationService`.

- **Behavior-identical structural refactor.** Same firing conditions, order, and user-facing strings. The big extraction was reviewed line-by-line.
- Completion streaming (`CompletionStrategy` / `StreamEvent` / renderer) untouched.
- Preserves #275's `reasoning_parts` on the assistant history message (re-applied into `runIteration`), with coverage.
- The observer seam is the substrate the planned **client-server split (#5)** will consume — a server implements `AgentLoopObserver` to serialize lifecycle events over the wire.

## 2. Default-persona capability reframe (#2)

Addresses the "same model feels a category below Codex on non-coding tasks" gap, diagnosed as prompt framing (not reasoning effort). In `personas/default/persona.md` (the single source of truth for the default prompt after #271):

- Adds a **Personality & Communication** section — be a thoughtful collaborator, match tone, lead with the answer, *reason and show thinking on analytical/open-ended requests*, state uncertainty.
- **Softens the blanket "action-first" line** — action-oriented for concrete operational tasks, but thinking/explaining for analytical/advisory/open-ended ones, instead of "Default to executing, not describing."

This is an **intentional behavior change** (that's the point). No doctrine removed.

## Tests

Full suite **1411 passing / 0 fail**, typecheck + lint + build clean. New: `agent-loop-observer.test.ts` (adapter mapping with byte-identical strings) and observer/reasoning/finalize cases in `agent-loop.test.ts`.

### Edge cases to verify in a staging session

1. User interrupt mid-generation → identical "generation stopped by user" behavior.
2. Iteration-limit hit → same "iteration limit reached (N)" warning, `finished=false`.
3. A non-coding/analytical prompt → the default agent now reasons/explains rather than rushing to a tool call.

## Notes for reviewers

- Part 1 is behavior-identical; Part 2 deliberately changes default-agent behavior via `persona.md`.
- Deferred (non-blocking) follow-ups: agent-loop branch coverage (meltdown / budget-pressure / compaction / toolsDisabled), an `agent-loop.test.ts` case that makes a live models.dev call, and optional prompt-fragment gating/stable-prefix (lower-value structural polish).

### CODE REVIEW

File: src/core/agent/execution/agent-loop.ts [hunk 1]
```diff
@@ -138,6 +165,447 @@ export interface CompletionStrategy {
...
+      totalToolsUsed: runMetrics.toolCalls,
+    });
+
+    // If the model produced reasoning but no text content (e.g. llama.cpp
+    // with --jinja routing the entire response into reasoning_content),
+    // surface the reasoning as the visible content so downstream
+    // consumers — conversation history, summarization, batch rendering —
+    // see what the model actually said.
+    const visibleContent = completion.content?.trim().length
+      ? completion.content
```
[the remark this answers is not in the document]
Comment (lvndry) [reply src/core/agent/execution/agent-loop.ts:594]: Declining: `ChatCompletionResponse.content` is typed `string` (non-optional), so `completion.content?.trim()` always yields a string and `.length` can't throw — the reasoning-only case is an empty string (`""`), which is falsy and correctly falls through to `reasoning`. The `?.` is redundant but harmless; there's no crash path here. (This code is a verbatim relocation from the pre-refactor loop, so behavior is unchanged.)

File: src/core/agent/execution/agent-loop.ts [hunk 2]
```diff
@@ -138,6 +165,447 @@ export interface CompletionStrategy {
...
+    });
+
+    for (const toolCall of toolCalls) {
+      if (toolCall.type === "function") {
+        state.recentToolCalls.push({
+          name: toolCall.function.name,
+          arguments: toolCall.function.arguments,
+        });
+      }
+    }
```
[the remark this answers is not in the document]
Comment (lvndry) [reply src/core/agent/execution/agent-loop.ts:291]: Declining: `ToolCall` has `type: "function"` as its sole variant and `function` as a **required** field (`{ name: string; arguments: string }`), so within the `type === "function"` branch `tc.function` is always defined per the types. These guards defend a state the type system disallows, and applying them would trade a hard crash for silently processing a malformed call as `"unknown"` — masking an upstream bug rather than surfacing it. If runtime-malformed tool calls from small local models are a real concern, the correct fix is one-time shape validation where `ToolCall`s are constructed in the LLM adapter, not scattered `?? "unknown"` guards in the loop. Out of scope for this behavior-identical refactor. (The access is a verbatim relocation from the pre-refactor loop.)

File: src/core/agent/execution/agent-loop.ts [hunk 3]
```diff
@@ -138,6 +165,447 @@ export interface CompletionStrategy {
...
+          state.currentMessages.push({
+            role: "tool",
+            name: toolCall.function.name,
+            content: formattedResult,
+            tool_call_id: toolCall.id,
+          });
+          recordToolResultTokens(runMetrics, toolCall.function.name, formattedResult.length);
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (lvndry) [reply src/core/agent/execution/agent-loop.ts:392]: Declining: `ToolCall` has `type: "function"` as its sole variant and `function` as a **required** field (`{ name: string; arguments: string }`), so within the `type === "function"` branch `tc.function` is always defined per the types. These guards defend a state the type system disallows, and applying them would trade a hard crash for silently processing a malformed call as `"unknown"` — masking an upstream bug rather than surfacing it. If runtime-malformed tool calls from small local models are a real concern, the correct fix is one-time shape validation where `ToolCall`s are constructed in the LLM adapter, not scattered `?? "unknown"` guards in the loop. Out of scope for this behavior-identical refactor. (The access is a verbatim relocation from the pre-refactor loop.)

File: src/core/agent/execution/agent-loop.ts [hunk 4]
```diff
@@ -138,6 +165,447 @@ export interface CompletionStrategy {
...
+        ? {
+            tool_calls: completion.toolCalls.map((tc) => ({
+              id: tc.id,
+              type: tc.type,
+              function: { name: tc.function.name, arguments: tc.function.arguments },
+              ...(tc.thought_signature ? { thought_signature: tc.thought_signature } : {}),
+            })),
+          }
+        : {}),
+    };
```
[the remark this answers is not in the document]
Comment (lvndry) [reply src/core/agent/execution/agent-loop.ts:562]: Declining: `ToolCall` has `type: "function"` as its sole variant and `function` as a **required** field (`{ name: string; arguments: string }`), so within the `type === "function"` branch `tc.function` is always defined per the types. These guards defend a state the type system disallows, and applying them would trade a hard crash for silently processing a malformed call as `"unknown"` — masking an upstream bug rather than surfacing it. If runtime-malformed tool calls from small local models are a real concern, the correct fix is one-time shape validation where `ToolCall`s are constructed in the LLM adapter, not scattered `?? "unknown"` guards in the loop. Out of scope for this behavior-identical refactor. (The access is a verbatim relocation from the pre-refactor loop.)

File: src/core/agent/agent-prompt.ts [hunk 5]
```diff
@@ -79,6 +85,18 @@ function getSkillIndexLineFromOption(s: {
...
+ * Whether a resolved persona is the built-in default persona, which gets
+ * fragment assembly with gating instead of the monolithic-string treatment
+ * used for custom personas. Checked both by the requested identifier (the
+ * common case) and by content equality with `DEFAULT_PROMPT` (covers the
+ * fallback path, where `resolvePersona` returns `FALLBACK_DEFAULT` for any
+ * unresolvable name).
+ */
+function isDefaultPersona(personaName: string, persona: AgentPersona): boolean {
+  return personaName.toLowerCase() === "default" || persona.systemPrompt === DEFAULT_PROMPT;
+}
```
[the remark this answers is not in the document]
Comment (lvndry) [reply src/core/agent/agent-prompt.ts]: Stale — this is anchored to commit `095596d`, which was part of the prompt-fragment approach (`DEFAULT_PROMPT`/`isDefaultPersona` in `system.ts`) that was **dropped** from this PR after #271 deleted `system.ts` and made `personas/default/persona.md` the single source of truth. Neither `isDefaultPersona` nor `DEFAULT_PROMPT` exists on the current branch, so there's nothing to change here. Resolving.
